hudi-io/src/main/java/org/apache/hudi/io/storage/HoodieStorage.java

```diff
@@ -0,0 +1,326 @@
+   *
+   * @param location given location.
+   * @param filter   filter to apply.
+   * @return the statuses of the files/directories in the given location.
+   * @throws FileNotFoundException when the location does not exist.
+   * @throws IOException           IO error.
+   */
+  List<HoodieFileStatus> listDirectEntries(HoodieLocation location,
+                                           HoodieLocationFilter filter) throws IOException;
+
+  /**
+   * Returns all the files that match the locationPattern and are not checksum files,
+   * and filters the results.
+   *
+   * @param locationPattern given pattern.
+   * @param filter          filter to apply.
+   * @return the statuses of the files.
+   * @throws IOException IO error.
+   */
+  List<HoodieFileStatus> globEntries(HoodieLocation locationPattern, HoodieLocationFilter filter)
+      throws IOException;
+
+  /**
+   * Renames the location from old to new.
+   *
+   * @param oldLocation source location.
+   * @param newLocation destination location.
+   * @return {@true} if rename is successful.
+   * @throws IOException IO error.
+   */
+  boolean rename(HoodieLocation oldLocation, HoodieLocation newLocation) throws IOException;
+
+  /**
+   * Deletes a directory at location.
+   *
+   * @param location directory to delete.
+   * @return {@code true} if successful.
+   * @throws IOException IO error.
+   */
+  boolean deleteDirectory(HoodieLocation location) throws IOException;
+
+  /**
+   * Deletes a file at location.
+   *
+   * @param location file to delete.
+   * @return {@code true} if successful.
+   * @throws IOException IO error.
+   */
+  boolean deleteFile(HoodieLocation location) throws IOException;
+
+  /**
+   * Qualifies a path to one which uses this storage and, if relative, made absolute.
+   *
+   * @param location to qualify.
+   * @return Qualified location.
+   */
+  HoodieLocation makeQualified(HoodieLocation location);
+
+  /**
+   * @return the underlying file system instance if exists.
+   */
+  Object getFileSystem();
+
+  /**
+   * @return the underlying configuration instance if exists.
+   */
+  Object getConf();
+
+  /**
+   * Create an OutputStream at the indicated location.
+   * The file is overwritten by default.
+   *
+   * @param location the file to create.
+   * @return the OutputStream to write to.
+   * @throws IOException IO error.
+   */
+  default OutputStream create(HoodieLocation location) throws IOException {
+    return create(location, true);
+  }
+
+  /**
+   * Creates a new file with overwrite set to false. This ensures files are created
+   * only once and never rewritten, also, here we take care if the content is not
+   * empty, will first write the content to a temp file if {needCreateTempFile} is
+   * true, and then rename it back after the content is written.
+   *
+   * @param location file Path.
+   * @param content  content to be stored.
+   */
+  default void createImmutableFileInPath(HoodieLocation location,
```

Comment:
We can make it final if non overwrite is expected.

Reply:
Makes sense.  Fixed.